### PR TITLE
Add timestamps to tooltips

### PR DESCRIPTION
**Pull request by Raader**
## Issue
#4488 
## Description
Adding timestamps to all tooltips
### Preview
Screen Shot 2022-08-27 at 18 25 53
Screen Shot 2022-08-27 at 18 25 57
Screen Shot 2022-08-27 at 18 26 05
Screen Shot 2022-08-27 at 18 26 09
Screen Shot 2022-08-27 at 21 06 48
Screen Shot 2022-08-27 at 22 53 42

### CODE REVIEW

File: web/src/components/tooltips/countrypanelexchangetooltip.jsx [hunk 1]
```diff
@@ -62,6 +69,7 @@ const CountryPanelExchangeTooltip = ({
 
   return (
     <Tooltip id="countrypanel-exchange-tooltip" position={position} onClose={onClose}>
+      <StyledTimeDisplay date={zoneData.stateDatetime ? new Date(zoneData.stateDatetime) : null} />
```
Comment (VIKTORVAV99) [inline web/src/components/tooltips/countrypanelexchangetooltip.jsx:72]:
```suggestion
      <StyledTimeDisplay date={zoneData.stateDatetime} />
```

File: web/src/components/tooltips/countrypanelproductiontooltip.jsx [hunk 2]
```diff
@@ -69,6 +76,7 @@ const CountryPanelProductionTooltip = ({ displayByEmissions, mode, position, zon
 
   return (
     <Tooltip id="countrypanel-production-tooltip" position={position} onClose={onClose}>
+      <StyledTimeDisplay date={zoneData.stateDatetime ? new Date(zoneData.stateDatetime) : null} />
```
Comment (VIKTORVAV99) [inline web/src/components/tooltips/countrypanelproductiontooltip.jsx:79]:
```suggestion
      <StyledTimeDisplay date={zoneData.stateDatetime} />
```

File: web/src/components/tooltips/pricetooltip.jsx [hunk 3]
```diff
@@ -14,6 +21,7 @@ const PriceTooltip = ({ position, zoneData, onClose }) => {
 
   return (
     <Tooltip id="price-tooltip" position={position} onClose={onClose}>
+      <StyledTimeDisplay date={zoneData.stateDatetime ? new Date(zoneData.stateDatetime) : null} />
```
Comment (VIKTORVAV99) [inline web/src/components/tooltips/pricetooltip.jsx:24]:
```suggestion
      <StyledTimeDisplay date={zoneData.stateDatetime} />
```

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
Overall this looks good but I added a few suggestions that should be addressed. Since we know the zone data will always exist then these views are shown and that it should always have a datetime we can remove the redundant checks for it.

I also think it would be great if you could update the carbon intensity graph in this PR as well so it also has the date time on top like the other graphs.

Currently it is below the main title as seen here:
image

**Comment by Raader** [comment]
@VIKTORVAV99 updated carbon intensity graph.

File: web/src/components/tooltips/mapcountrytooltip.jsx [hunk 4]
```diff
@@ -92,9 +92,9 @@ const MapCountryTooltip = ({ electricityMixMode, position, zoneData, onClose })
 
   return (
     <Tooltip id="country-tooltip" position={position} onClose={onClose}>
+      <StyledTimeDisplay date={zoneData.stateDatetime ? new Date(zoneData.stateDatetime) : null} />
```
Comment (VIKTORVAV99) [inline web/src/components/tooltips/mapcountrytooltip.jsx:95]:
```suggestion
      <StyledTimeDisplay date={zoneData.stateDatetime} />
```

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
Just one more suggestion then I looks ready to me.

**Review by VIKTORVAV99 — APPROVED** [review]
LGTM!

@rasmusjeppsen what do you think, any design feedback that needs to be addressed?

@Kongkille am I right in assuming every datapoint will always have a datetime object and the checks aren't needed?

**Comment by rasmusjeppsen** [comment]
> @rasmusjeppsen what do you think, any design feedback that needs to be addressed?

Thank you for providing a PR for this, @Raader! Looks great. And thank you for your input/review @VIKTORVAV99. Much appreciated!

In terms of the design, I think increasing the visibility of the timestamp by making the text bold would be a nice detail to add.
Here are some design mockups based on the screenshots @Raader provided in the PR to illustrate:

timestamp_mockup
timestamp_mockup2

> I also think it would be great if you could update the carbon intensity graph in this PR as well so it also has the date time on top like the other graphs.

Nice catch, Viktor! I agree - this would be great for consistency.

One thing to note here is that we show the country first and then the timestamp second when hovering on the map zones (outside the sidepanel graphs), which then arguably leads to some inconsistency of the sidepanel vs. the map.

zone_hover_example

I do, however, think that in terms of the hierarchy of information displayed, the country is the essential information when navigating/browsing the map, whereas the timestamp is the essential information for the graphs on the sidepanel.
So, **TLDR;** I agree with updating the carbon intensity graph, but think we should keep the layout of having the country first and timestamp second when hovering over the zones on the map. We could address this in a future PR to improve overall consistency if need be.

Great work!

**Comment by Raader** [comment]
@rasmusjeppsen even though I think making timestamps bold will make it look better, wouldn't it introduce even greater inconsistency between the map tooltips and the sidebar tooltips. Updating the map tooltips appearance is might be out of scope for this PR.

**Comment by rasmusjeppsen** [comment]
> @rasmusjeppsen even though I think making timestamps bold will make it look better, wouldn't it introduce even greater inconsistency between the map tooltips and the sidebar tooltips. Updating the map tooltips appearance might be out of scope for this PR.

@Raader I understand the concern regarding inconsistency, however, I think making the timestamps bold in the tooltips for the sidepanel is the right step forward here whilst keeping the following position of the timestamp for the sidepanel as per your preview:
image

And this position for the map as it currently is:
image

I also think that making the timestamps bold in the side panel will further differentiate the layout of tooltips in the sidepanel vs. when hovering on zones on the map which IMO makes the two layouts prioritize the display of information differently (timestamp vs. country first), rather than making it inconsistent in a bad/unthoughtful way.

I suggest we add the bold text to the timestamps in the sidepanel and then everything LGTM!

**Comment by Raader** [comment]
@rasmusjeppsen 
> I also think that making the timestamps bold in the side panel will further differentiate the layout of tooltips in the sidepanel vs. when hovering on zones on the map which IMO makes the two layouts prioritize the display of information differently (timestamp vs. country first), rather than making it inconsistent in a bad/unthoughtful way.
> 
Makes sense. Made them bold.

**Comment by VIKTORVAV99** [comment]
> > @Kongkille am I right in assuming every datapoint will always have a datetime object and the checks aren't needed?
> 
> @VIKTORVAV99 @Kongkille Map country tooltip(the one which had timestamps already) originally had a date time check. Maybe we should've kept those checks just in case.

I think it should be safe as the backend fill the dates for empty values and the graphs should not be visible at all without this data but we'll see what Kongkille says. And even if there is checks needed we could simplify them with optional chaining anyway.

**Comment by Kongkille** [comment]
> @Kongkille am I right in assuming every datapoint will always have a datetime object and the checks aren't needed?

Yes that should always be the case

**Review by Kongkille — COMMENTED** [review]
Looks great!

On the code side of things, I'm seeing a bit of duplication in the styling, it's not a big deal but the timedisplay could have a default styling corresponding to the shared styling and then we can override for edge cases.

You could do that by styling the <p> tag in timeDisplay.jsx

On UI, I think it looks great and is intuitive. Have you considered adding some kind of clock icon next to the date? 

image

**Comment by VIKTORVAV99** [comment]
> Looks great!
> 
> On the code side of things, I'm seeing a bit of duplication in the styling, it's not a big deal but the timedisplay could have a default styling corresponding to the shared styling and then we can override for edge cases.
> 
> You could do that by styling the
> 
> tag in timeDisplay.jsx
> 
> On UI, I think it looks great and is intuitive. Have you considered adding some kind of clock icon next to the date?
> 
> image

Adding a icon is a good idea, but I think maybe a calendar icon would be a better fit?
Or switching between a clock icon for "live" view and a calendar icon for historical view.
Calendar month icon:
calendar_month_FILL0_wght400_GRAD0_opsz48

Or perhaps the date range or today icon:
date_range_FILL0_wght400_GRAD0_opsz48
today_FILL0_wght400_GRAD0_opsz48

Or a combination of date range/calendar month, today, and the clock depending on the selected time resolution? (If this is believed to be outside the scope of the PR I can create a subsequent PR with the icon changes.)

**Comment by Raader** [comment]
> Or a combination of date range/calendar month, today, and the clock depending on the selected time resolution? (If this is believed to be outside the scope of the PR I can create a subsequent PR with the icon changes.)

@VIKTORVAV99 I think you should open a subsequent PR.

**Review by Kongkille — APPROVED** [review]
I think this is good to go!

Thanks a lot @Raader for driving this through

**Comment by Raader** [comment]
@Kongkille is there something that keeps this from being merged?

**Comment by VIKTORVAV99** [comment]
> @Kongkille is there something that keeps this from being merged?

Kongkille is on vacation but I don't think there is anything blocking it.
@tonypls what do you think, we should be able to merge this right?

**Comment by tonypls** [comment]
> > @Kongkille is there something that keeps this from being merged?
> 
> Kongkille is on vacation but I don't think there is anything blocking it. @tonypls what do you think, we should be able to merge this right?

Yes this is good to go! 🚀 

@Raader Feel free to hit the merge button when the tests have passed 👌

**Comment by tonypls** [comment]
Ah okay that makes sense 😄  I'll merge it for ya. Nice work!

**Pull request merged**